apps/emqx/src/emqx_cluster_rpc.erl

```diff
@@ -0,0 +1,322 @@
+                _ -> {next_state, ?CATCH_UP, Node, [catch_up_after(1)]}
+            end;
+        {aborted, _Reason} -> {next_state, ?CATCH_UP, Node, [catch_up_after(1)]}
+    end.
+
+get_next_mfa(Node) ->
+    Fun =
+        fun() ->
+            case mnesia:read(?CLUSTER_CURSOR, Node, write) of
+                [] ->
+                    LatestId = get_latest_id(),
+                    Cursor = #cluster_rpc_cursor{node = Node, tnx_id = max(LatestId - 1, 0)},
+                    mnesia:write(?CLUSTER_CURSOR, Cursor, write),
+                    case mnesia:read(?CLUSTER_MFA, LatestId) of
+                        [] -> caught_up;
+                        [#cluster_rpc_mfa{mfa = MFA}] -> {still_lagging, LatestId, MFA}
+                    end;
+                [#cluster_rpc_cursor{tnx_id = DoneTnxId}] ->
+                    case mnesia:read(?CLUSTER_MFA, DoneTnxId + 1) of
+                        [] -> caught_up;
+                        [#cluster_rpc_mfa{mfa = MFA}] -> {still_lagging, DoneTnxId + 1, MFA}
+                    end
+            end
+        end,
+    transaction(Fun).
+
+do_catch_up(ToTnxId, Node) ->
+    case mnesia:wread({?CLUSTER_CURSOR, Node}) of
+        [] ->
+            Rec = #cluster_rpc_cursor{tnx_id = ToTnxId, node = Node},
+            mnesia:write(?CLUSTER_CURSOR, Rec, write),
+            caught_up;
+        [#cluster_rpc_cursor{tnx_id = DoneTnxId}] when ToTnxId =< DoneTnxId -> caught_up;
+        [Rec = #cluster_rpc_cursor{tnx_id = DoneTnxId}] ->
+            CurTnxId = DoneTnxId + 1,
+            [#cluster_rpc_mfa{mfa = MFA}] = mnesia:read(?CLUSTER_MFA, CurTnxId),
+            mnesia:write(?CLUSTER_CURSOR, Rec#cluster_rpc_cursor{tnx_id = CurTnxId}, write),
+            apply_mfa(CurTnxId, MFA)
+    end.
+
+get_latest_id() ->
+    case mnesia:last(?CLUSTER_MFA) of
+        '$end_of_table' -> 0;
+        Id -> Id
+    end.
+
+handle_mfa_write_event(#cluster_rpc_mfa{tnx_id = TnxId, mfa = MFA}, Node) ->
+    X = transaction(fun() -> get_done_id(Node, TnxId - 1) end),
+    case X of
+        {atomic, DoneTnxId} when DoneTnxId =:= TnxId - 1 ->
+            case apply_mfa(TnxId, MFA) of
+                ok ->
+                    Trans = fun() ->
+                        Rec = #cluster_rpc_cursor{tnx_id = TnxId, node = Node},
+                        mnesia:write(?CLUSTER_CURSOR, Rec, write)
+                            end,
+                    case transaction(Trans) of
+                        {atomic, ok} ->
+                            ?LOG(warning, "ooooo~p", [del_mfa_after(2*50)]),
+                            {next_state, ?IDLE, Node, [del_mfa_after(1 * 60)]};
+                        _ -> {next_state, ?CATCH_UP, Node, [catch_up_after(1)]}
+                    end;
+                _ -> {next_state, ?CATCH_UP, Node, [catch_up_after(1)]}
+            end;
+        {atomic, _DoneTnxId} -> {next_state, ?CATCH_UP, Node, [catch_up_after(0)]};
+        _ -> {next_state, ?CATCH_UP, Node, [catch_up_after(1)]}
```

Comment:
this is not going to happen ?

Reply:
if mnesia is broken, we don't let this process crash, catch up later.

Reply:
Could you elaborate with an example of “mnesia is broken” e.g in which way

Reply:
I just think the transaction can return aborted for some reason.
if aborted happened , we can just crash and restart.
```erlang
{atomic, DoneTnxId} = transaction(fun() -> get_done_id(Node, TnxId - 1) end).
```